operators/test/e2e/apm/association_test.go

```diff
@@ -0,0 +1,112 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package apm
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/elastic/cloud-on-k8s/operators/pkg/apis/apm/v1alpha1"
+	commonv1alpha1 "github.com/elastic/cloud-on-k8s/operators/pkg/apis/common/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/annotation"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/events"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/operators/test/e2e/test"
+	"github.com/elastic/cloud-on-k8s/operators/test/e2e/test/apmserver"
+	"github.com/elastic/cloud-on-k8s/operators/test/e2e/test/elasticsearch"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func TestAPMAssociationWithNonExistentES(t *testing.T) {
```

Comment:
Is this something we could catch with unit tests directly?
The more E2E tests the better, probably, but if we introduce E2E tests every time we emit an event, that's quite a lot of clusters to spawn on our testing infra.

(Totally fine with keeping things in this PR, it's more like something we should think about for the long term)

Reply:
I wanted to write a unit test as well but I couldn't find any existing ones for the full reconciliation loop. Because it would take some time, I felt that writing that test is probably better done in a different PR.